Add initial version of Vilfo Router integration

ManneW opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
New integration added for the Vilfo Router (https://www.vilfo.com). The Vilfo Router is a VPN router with an HTTP API. In this initial version, the integration only provides a couple of sensors to monitor the status of the router. The idea is to add device tracking soon after this initial version (but as the guidelines states, this first version is limited to a single platform).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:

Integration only supports config_flow configuration.

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- The currently supported sensors are router uptime and router load.
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/11868

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)

For the `config_flow` part of the integration:
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [X] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] Untested files have been added to `.coveragerc`.

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

<!--

This commit is a combination of several commits, with commit messages in bullet form below.

* Added additiional files to Vilfo integration.
* Added generated files.
* Fixed alphabetic order in generated config_flows.
* Continued implementation of config flow for Vilfo integration.
* Continued work on config_flow for Vilfo.
* Updated requirements in manifest for Vilfo Router integration.
* Some strings added to Vilfo Router integration.
* Vilfo Router integration updated with sensor support.
* Code style cleanup.
* Additional cleanup of config flow.
* Added additional UI strings for Vilfo Router
* Updated tests of config flow and fixed formatting
* Updated requirement upon vilfo-api-client.
* Sensor refactoring including support for icons
* Code style changes for Vilfo Router integration
* Code cleanup

-->

ManneW committed:
Initial implementation of Vilfo router integration.

This commit is a combination of several commits, with commit messages in bullet form below.

* Added additional files to Vilfo integration.
* Added generated files.
* Fixed alphabetic order in generated config_flows.
* Continued implementation of config flow for Vilfo integration.
* Continued work on config_flow for Vilfo.
* Updated requirements in manifest for Vilfo Router integration.
* Some strings added to Vilfo Router integration.
* Vilfo Router integration updated with sensor support.
* Code style cleanup.
* Additional cleanup of config flow.
* Added additional UI strings for Vilfo Router
* Updated tests of config flow and fixed formatting
* Updated requirement upon vilfo-api-client.
* Sensor refactoring including support for icons
* Code style changes for Vilfo Router integration
* Code cleanup
* Fixed linting issues in Vilfo Router integration
* Fixed import order in test for Vilfo integration.

homeassistant commented:
Hello @ManneW,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`e846ff2770b3b0f000a161fc2bad9c8c5f3f357f`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;home-assistant&#x2F;pull&#x2F;31177&#x2F;commits&#x2F;e846ff2770b3b0f000a161fc2bad9c8c5f3f357f) - This commit has something that looks like an email address (emanuel.winblad@sportsediting.se). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @ManneW, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31177).

Once you do that we will be able to review and accept this pull request.

Thanks!

Quentame commented:
Sorry for the spam I made a mistake

ManneW commented:
No worries! Were you trying to fix the `cla-error` label? I think that it can be removed now anyhow 🙂

Quentame commented:
Yeppa, I don't know why you have this now (also I was on a mobile app, and when I've add `cla-recheck` removed all other tags).

The recheck is not removing it. I'll let people who knows taking take of it.

Or maybe "ManneW authored and Emanuel Winblad committed" on the first commit, try rebase and force push, so it will be you the author and committer.

So, I'll review soon :blush:

ManneW commented:
I actually did amend the "author" part to resolve the CLA issue from the beginning (before that it was only "Emanuel Winblad" and not "ManneW") 😄 

But your suggestion was a good idea, so I did the rebase now to have things a bit cleaner 🙂Not sure if the `cla-error` label will go away though, but we'll see.

Quentame commented on homeassistant/components/vilfo/sensor.py:
```diff
@@ -0,0 +1,105 @@
+    name = ATTR_ROUTER_NAME
+    device_info = {
+        "identifiers": {(DOMAIN, vilfo.host, vilfo.mac_address)},
+        "name": name,
```
```suggestion
        "name": ATTR_ROUTER_NAME,
```

and remove name var

Quentame commented on homeassistant/components/vilfo/sensor.py:
```diff
@@ -0,0 +1,105 @@
+        "identifiers": {(DOMAIN, vilfo.host, vilfo.mac_address)},
+        "name": name,
+        "manufacturer": ATTR_ROUTER_MANUFACTURER,
+        "model": ATTR_ROUTER_MODEL,
```
A name is different from the model.

And `ATTR_ROUTER_NAME` == `ATTR_ROUTER_MODEL` here.

The model should be the "hardware version", remove it here.

ManneW replied:
Ah, ok! I think I understand the difference.

In this case, I think there is only one hardware version available of the router (to this date).  Would it be better to change the name into something else or more generic? Like within this integration, the router is basically that -- "Router".

I'm not sure what's the best practice when it comes to device tracker and connected devices, but Vilfo also has the concept of "VPN groups" which can have connection statuses etc. Would these "VPN groups" be suitable to add as additional entities later on (as children to the router).

Just trying to wrap my head around the concepts here 😄

ManneW replied:
@Quentame I gave this some more thought (regarding the `name` being the same as the `model`).

Since the `name` can be overridden/customized by the user, it think it makes sense to have the `model` indicating that it is indeed a `Vilfo Router` and nothing else. Does that sound reasonable? 🙂 

Do you have any input on this, @MartinHjelmare ?

MartinHjelmare replied:
The items here should represent the device and not the entity. If it makes sense for the device, it's good. I don't know the device so I can't give specific advice.

Quentame commented on homeassistant/components/vilfo/const.py:
```diff
@@ -0,0 +1,31 @@
+
+
+SENSOR_TYPES = {
+    "generic": {ATTR_LABEL: ATTR_GENERIC.title(), ATTR_UNIT: None, ATTR_ICON: None},
```
The generic sensor isn't used, and don't think it will.
Remove this line.

Quentame commented on homeassistant/components/vilfo/sensor.py:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    sensors.append(VilfoRouterLoadSensor(vilfo, device_info))
+    sensors.append(VilfoRouterUptimeSensor(vilfo, device_info))
```
IF you want:

- You can remove your non generic sensor classes
- Add a `type` param to  `VilfoRouterSensor`
- Foreach `SENSOR_TYPES` and pass the key to the `type param.`

ManneW replied:
Great suggestion! I did that along with some other refactoring which made it much more clear and to the point 👍 😄

Quentame commented on tests/components/vilfo/test_config_flow.py:
```diff
@@ -0,0 +1,81 @@
+            {CONF_HOST: "testadmin.vilfo.com", CONF_ACCESS_TOKEN: "test-token"},
+        )
+
+    assert result2["type"] == "create_entry"
```
`from homeassistant import data_entry_flow`

create_entry --> `data_entry_flow.RESULT_TYPE_CREATE_ENTRY`
form --> `data_entry_flow.RESULT_TYPE_FORM`

ManneW replied:
I noticed now that I missed this one. Will go over the tests again soon and make some updates.

ManneW replied:
Done! ✅

Quentame commented on homeassistant/components/vilfo/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
```
Useless log, might be more interested : 
```suggestion
    except Exception as err:
                _LOGGER.error("Unexpected exception: %s", err)
```

ManneW replied:
:+1:

Quentame reviewed (changes requested):
Why there is a `src/vilfo-api-client` submodule here ?
It shouldn't be.

Quentame commented:
`config_flow.py` needs 100% code coverage

ManneW commented on homeassistant/components/vilfo/__init__.py:
```diff
@@ -0,0 +1,114 @@
+        """Get the unique_id for the Vilfo Router."""
+        if self.mac_address:
+            return self.mac_address
+        return self.host
```
I can have a MAC address, but since the resolution of it's MAC address (in the `vilfo-api-client`) relies on `get-mac` (https://pypi.org/project/get-mac/) there might be situations where it's not possible to resolve the router MAC. Hence, I thought it would be good to use it if available, but otherwise don't rely on it. But I might have thought wrong 😄

MartinHjelmare replied:
An IP address is not allowed as unique id as it may change:
https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements

It's ok to use the mac address when that's available, otherwise fall back to `None`.

ManneW replied:
Ok :+1:

As the variable name implies, it can also be a hostname (`admin.vilfo.com` being the default for such a router). I guess a hostname would be fine to use as unique ID? :smile:

MartinHjelmare replied:
If it doesn't change a hostname may be ok.

ManneW replied:
@MartinHjelmare I've changed it to only use the `host` as `unique_id` when the `host` is set to `admin.vilfo.com`, which is the default one and that isn't possible to be modified (at least not supported by Vilfo). I've pushed these changes to update the PR.

ManneW committed:
Updates to Vilfo Router integration based on feedback

Based on the feedback received, updates have been made to the Vilfo Router integration.

A couple of the points mentioned have not been addressed yet, since the appropriate action has not yet been determined. These are:

* https://github.com/home-assistant/home-assistant/pull/31177#discussion_r371124477
* https://github.com/home-assistant/home-assistant/pull/31177#discussion_r371202896

This commit consists of:

    * Removed unused folder/submodule
    * Fixes to __init__
    * Fixes to config_flow
    * Fixes to const
    * Refactored sensors and applied fixes
    * Fix issue with wrong exception type in config flow
    * Updated tests for Vilfo integration config_flow
    * Updated dependency upon vilfo-api-client to improve testability
    * Import order fixes in test
    * Use constants instead of strings in tests

ManneW commented on homeassistant/components/vilfo/config_flow.py:
```diff
@@ -0,0 +1,124 @@
+
+                return self.async_create_entry(title=info["title"], data=user_input)
+            except AlreadyConfigured:
+                return self.async_abort(reason="already_configured")
```
If I'm reading that method correctly, you're absolutely right!

ManneW commented:
Don't want to be pushy, but do you think you will have time to check out the comments and changes anytime soon, @Quentame? :smile:

ManneW committed:
Updated the VilfoRouterData class to only use the hostname as unique_id when it is the default one (admin.vilfo.com).

ManneW commented on homeassistant/components/vilfo/__init__.py:
```diff
@@ -0,0 +1,120 @@
+            boot_time = dt_util.parse_datetime(board_information["bootTime"])
+            uptime = dt_util.now() - boot_time
+            uptime_seconds = round(uptime.total_seconds(), 0)
+            uptime_minutes = round(uptime_seconds / 60, 0)
```
I didn't know about the `timestamp` device class, which seems to come in handy here and make it clearer, so I'll change to that :+1:

Perhaps I should rename it from `uptime` to `boot_time` or something as well? I guess the `timestamp` device class is to mark a point in time, rather than a duration? 🕥

ManneW replied:
When I went about to refactor this, I noticed that these relative times are never computed in the sensors themselves, but only during update (which is throttled). I think this should basically be the same as if the API library was doing the calculation. Not sure if it makes any difference in this case? Maybe it still should be avoided to keep things uniform 🙂

MartinHjelmare replied:
Yes, please use the timestamp device class and absolute time as state.

MartinHjelmare replied:
The lovelace frontend is able to show the absolute timestamp as a relative time.

https://www.home-assistant.io/lovelace/entities/#format

ManneW committed:
Refactored based on feedback during review.

* Changes to constant names,
* Blocking IO separated to executor job,
* Data for uptime sensor changed from being computed to being a timestamp,
* Started refactoring uptime sensor in terms of naming and unit.
* Updated constants for boot time (previously uptime) sensor.
* Refactored test of Vilfo config flow to avoid patching code under test.
* UI naming fixes and better exception handling.

ManneW commented on homeassistant/components/vilfo/config_flow.py:
```diff
@@ -0,0 +1,121 @@
+    # This doesn't validate authentication.
+    controller = VilfoClient(host=data[CONF_HOST], token=data[CONF_ACCESS_TOKEN])
+    try:
+        controller.ping()
```
> Is this async safe? Is there blocking I/O here?

Ah, yeah. I think there might be 🤦‍♂ 

I've refactored that so that both `controller.ping()` and `controller.get_board_information()` are called in an executor job.

I will finalize an update to the PR tomorrow 🙂

ManneW committed:
Removed unused exception class.

MartinHjelmare commented on homeassistant/components/vilfo/const.py:
```diff
@@ -0,0 +1,38 @@
+    },
+    "boot_time": {
+        ATTR_LABEL: ATTR_BOOT_TIME.replace("_", " ").title(),
+        ATTR_UNIT: UNIT_ISO8601,
```
The timestamp device class sensor is unit less.

ManneW replied:
OK, I can remove that. I based it off the docs you linked to:

> If specifying a device class, your sensor entity will need to also return the correct unit of measurement.

In the table the "unit" for `timestamp` is `ISO8601` (https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes)

ManneW replied:
@MartinHjelmare Let me know your thoughts on how we can make that documentation better (regarding the `timestamp` and its unit) and I'll try to update that in a developer documentation PR that I'm about to start within a day or so. I've been trying to gather the things I've come across when using the docs 😄

MartinHjelmare replied:
Not sure how to make the table clear. Add a fourth column with comments and remove ISO.... from the unit column?

MartinHjelmare commented on homeassistant/components/vilfo/sensor.py:
```diff
@@ -0,0 +1,91 @@
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return SENSOR_TYPES[self.sensor_type].get(ATTR_DEVICE_CLASS, None)
```
`None` is the default value returned if the key is missing in the dict.

MartinHjelmare commented on tests/components/vilfo/test_config_flow.py:
```diff
@@ -0,0 +1,172 @@
+    )
+
+    with patch(
+        "homeassistant.components.vilfo.config_flow.validate_input",
```
Don't patch our code.

MartinHjelmare commented on tests/components/vilfo/test_config_flow.py:
```diff
@@ -0,0 +1,172 @@
+    )
+
+    with patch(
+        "homeassistant.components.vilfo.config_flow.validate_input",
```
See above.

MartinHjelmare commented on tests/components/vilfo/test_config_flow.py:
```diff
@@ -0,0 +1,172 @@
+    )
+
+    with patch(
+        "homeassistant.components.vilfo.config_flow.validate_input",
```
See above.

ManneW committed:
Various changes to Vilfo Router integration.

* Removed unit of measurement for boot time sensor,
* Added support for a sensor not having a unit,
* Updated the config_flow to handle when the integration is already configured,
* Updated tests to avoid mocking the code under test and also to cover the aforementioned changes.

MartinHjelmare commented on homeassistant/components/vilfo/config_flow.py:
```diff
@@ -0,0 +1,148 @@
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except data_entry_flow.AbortFlow as abort_err:
```
Why catch this and just reraise? Looks like we can remove this.

ManneW replied:
This is to prevent it from being caught by the "unknown error"-clause, which otherwise catches it.

ManneW replied:
Would a bare `raise` instead of `raise abort_err` be preferable?

MartinHjelmare replied:
We should move setting unique_id etc out of this try...except. The try...except block should only contain the line that we except to raise.

ManneW replied:
`else:` block added with the `unique_id` stuff in it. More readable and cleaner now :+1:

MartinHjelmare reviewed (commented):
See below.

Quentame commented:
Hi again @ManneW sorry for not beeing responsive, I am busy on iCloud issues :sweat_smile: 
But I saw you are on good hands !

Checked my old comments, all resolved :+1:

ManneW committed:
Exception handling in Vilfo Router config flow refactored to be more readable.

ManneW commented:
I've updated (but not pushed yet) this based on the last comments from @MartinHjelmare. Do you want me to squash all the work into a single commit?

I'll push it as a single new commit in the meantime, but are ready to do some squashing to keep the repo as clean as possible 😄

MartinHjelmare commented:
Please don't squash commits after review has started to make it easier for readers to track changes.

We'll squash when merging. :+1:

MartinHjelmare reviewed (approved):
Very nice!

MartinHjelmare commented:
Can be merged when build passes.

ManneW commented:
Thank you for great review feedback, @MartinHjelmare and @Quentame! 🙏 

No worries, @Quentame! 😄  I guess it would be good if you also marked your original review as "approved" (or if one can remove the review or something?) so that the PR can be sorted accordingly 🙂

MartinHjelmare commented on homeassistant/components/vilfo/__init__.py:
```diff
@@ -0,0 +1,114 @@
+    async def async_update(self):
+        """Update data using calls to VilfoClient library."""
+        try:
+            board_information = self._vilfo.get_board_information()
```
Does this do I/O?

ManneW replied:
Yeah, I believe it does. The `async_update` method is the only place where the API client library is called directly.

ManneW replied:
I'm doing a refactoring of this now.

ManneW replied:
I'm updating the PR right now with a proposed fix for this.

MartinHjelmare commented on homeassistant/components/vilfo/__init__.py:
```diff
@@ -0,0 +1,114 @@
+            board_information = self._vilfo.get_board_information()
+            self.firmware_version = board_information["version"]
+            self.data["boot_time"] = board_information["bootTime"]
+            self.data["load"] = self._vilfo.get_load()
```
And this?

ManneW replied:
This should be included in the other fix mentioned above 🙂

MartinHjelmare reviewed (changes requested):
Sorry, I think I've missed this.

Quentame commented on homeassistant/components/vilfo/const.py:
```diff
@@ -0,0 +1,36 @@
+        ATTR_API_DATA_FIELD: ATTR_API_DATA_FIELD_LOAD,
+    },
+    "boot_time": {
+        ATTR_LABEL: ATTR_BOOT_TIME.replace("_", " ").title(),
```
You don't need the `ATTR_LOAD` + `ATTR_BOOT_TIME` constants.

Just add clean labels here.

```suggestion
        ATTR_LABEL: "Boot time",
```

Quentame replied:
Or use those constants in init line 99,100 maybe more

Quentame commented on homeassistant/components/vilfo/sensor.py:
```diff
@@ -0,0 +1,91 @@
+        """Initialize."""
+        self.api = api
+        self.sensor_type = sensor_type
+        self._device_info = device_info
```
Initialize `device_info` here  or return directly in the property.

So you can remove the `device_info` param.

ManneW replied:
Sorry, but I don't follow 😄 

**Do you mean refactor into something like:**

```python
# ...
def __init__(self, sensor_type, api, device_info):
    # ...
    self.device_info = device_info
# ...
```
(and then remove the `device_info` method)

**or do you mean:**

Pass all the attributes of the `device_info` object (defined in `async_setup_entry`) as individual parameters to the constructor (and then stitching them together as a single object again)?

Quentame replied:
I mean : 

```suggestion
        self._device_info = {
        "identifiers": {(DOMAIN, vilfo.host, vilfo.mac_address)},
        "name": ROUTER_DEFAULT_NAME,
        "manufacturer": ROUTER_MANUFACTURER,
        "model": ROUTER_DEFAULT_MODEL,
        "sw_version": vilfo.firmware_version,
    }
```

Quentame commented on homeassistant/components/vilfo/sensor.py:
```diff
@@ -0,0 +1,91 @@
+        """Return the name of the sensor."""
+        parent_device_name = self._device_info["name"]
+        sensor_name = SENSOR_TYPES[self.sensor_type][ATTR_LABEL]
+        return f"{parent_device_name} {sensor_name}"
```
`parent_device_name` will always be the same (`ROUTER_DEFAULT_NAME`), not that unique ?

Did I miss something ?

ManneW replied:
If the user manually renames the parent device (the router in this case), the name will propagate to the child entities (if you haven't customized their respective names). See attached screenshots.

### Before renaming

![before_renaming](https://user-images.githubusercontent.com/659680/74342953-981c6c00-4daa-11ea-9967-6217460f2efb.png)

### After renaming

![after_renaming](https://user-images.githubusercontent.com/659680/74342956-98b50280-4daa-11ea-8434-06910217d36a.png)

Quentame replied:
Ok then, I missed something 😅

Resolving

Quentame reviewed (commented):
3 comments more before merging

ManneW committed:
Refactored constant usage, fixed sensor availability and fix API client library doing I/O in async context.

MartinHjelmare commented on homeassistant/components/vilfo/__init__.py:
```diff
@@ -69,9 +69,10 @@
     """Define an object to hold sensor data."""
 
-    def __init__(self, host, access_token):
+    def __init__(self, host, access_token, hass):
```
Please put `hass` first in the signature. That's our standard whenever we need to pass `hass`.

ManneW replied:
:+1: Done! I'll add that to that list of things to add to the developer documentation (at least I haven't seen it there) 😄

ManneW commented on homeassistant/components/vilfo/__init__.py:
```diff
@@ -69,9 +69,10 @@
     """Define an object to hold sensor data."""
 
-    def __init__(self, host, access_token):
+    def __init__(self, host, access_token, hass):
```
```suggestion
    def __init__(self, hass, host, access_token):
```

ManneW committed:
Updated signature with hass first

ManneW commented on homeassistant/components/vilfo/__init__.py:
```diff
@@ -33,7 +33,7 @@
     access_token = entry.data[CONF_ACCESS_TOKEN]
 
-    vilfo_router = VilfoRouterData(host, access_token)
+    vilfo_router = VilfoRouterData(host, access_token, hass)
```
```suggestion
    vilfo_router = VilfoRouterData(hass, host, access_token)
```

ManneW committed:
Update call to constructor with changed order of arguments

MartinHjelmare reviewed (approved):
Thanks!

Quentame commented:
Very nice work @ManneW !

Thanks for contributing 👏
